# 03 – Finaler Durchlauf: beam=12, len=20 über alle Sessions (vor Bugfix)

## Kontext

Die Experimente in `02_`–`02h_` wurden auf einem **5-Session-Dev-Subset** durchgeführt.
Dieses Notebook führt den bisher besten Parameter-Stand (beam=12, max_length=20)
auf **allen Dev-Sessions** aus, um eine robuste Gesamtschätzung zu erhalten.

**Hinweis:** Dieser Lauf verwendet noch die Daten **vor dem Segmentierungs-Bugfix**. Der Bugfix-Lauf folgt in `03b_`.

## Ablauf

1. Inference auf allen Sessions mit `inference.main()` (auskommentiert – läuft nicht erneut)
2. Evaluation via `evaluate.py` als Subprocess
3. Regex-Parsing der Evaluations-Ausgabe → DataFrame
4. Vergleich mit der ursprünglichen BL4-Baseline (beam=3, len=15)

## Ergebnis (Vorschau)

| Konfiguration | WER | Δ WER | Joint Error |
|--------------|-----|-------|-------------|
| Baseline BL4 (beam=3, len=15) | 0.4987 | – | 0.3457 |
| **Final (beam=12, len=20)** | **0.4980** | **−0.0008 (−0.15%)** | **0.3453** |

Marginale Verbesserung.

**Hinweis zum Bugfix:** Dieser finale Lauf basiert auf dem Stand **vor dem Segmentierungs-Bugfix**. Der Bugfix-Lauf folgt in `03b_`. Die hier erzeugten VTT-Dateien (`output_final_bs12_len20`) dienen als Eingabe für die LLM-Postprocessing-Notebooks `03a_` und `03a2_`.

## 1 – GPU-Check

In [9]:
!nvidia-smi --query-gpu=index,memory.total,memory.used,memory.free --format=csv,noheader

0, 81920 MiB, 23041 MiB, 57996 MiB
1, 81920 MiB, 4 MiB, 81033 MiB
2, 81920 MiB, 81018 MiB, 19 MiB
3, 81920 MiB, 73976 MiB, 7061 MiB


In [10]:
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "2"  # macht "unsere" GPU zur 0 im Prozess

import torch
print("n_gpu:", torch.cuda.device_count())
print("current device:", torch.cuda.current_device())
print("device name:", torch.cuda.get_device_name(0))


n_gpu: 1
current device: 0
device name: NVIDIA A100-SXM4-80GB


## 2 – Pfad-Konfiguration

`session = "*"` nutzt Shell-Glob – `inference.main()` verarbeitet alle Sessions
deren Verzeichnis auf dieses Muster passt.

In [11]:
project_baseline_path = "/home/josch080/Projektgruppe/mcorec_baseline" # anpassen
session = "*" # alle Sessions
session_dir = f"data-bin/dev/{session}"

In [12]:
import os, sys
os.chdir(project_baseline_path)
print("CWD:", os.getcwd())

CWD: /home/josch080/Projektgruppe/mcorec_baseline


In [13]:
from script import inference

/home/josch080/Projektgruppe/mcorec_baseline/script/inference.py:22: FutureWarning: In the future `np.object` will be defined as the corresponding NumPy scalar.
  if not hasattr(np, "object"):


## 3 – Inference-Parameter

In [14]:
model_type = "avsr_cocktail"
checkpoint_path = "model-bin/avsr_cocktail_mcorec_finetune" # BL4-Checkpoint
beam_size = 12 # beste Kombination aus 02_ Grid-Search
max_length = 20 # beste Kombination aus 02_ Grid-Search
output_dir_name = f"output_final_bs12_len20"

## 4 – Inference starten (auskommentiert)

`sys.argv` simuliert einen Kommandozeilenaufruf für `inference.main()`.
Auskommentiert – Inference-Ergebnisse liegen bereits vor.
`session_dir` mit `*`-Glob verarbeitet alle Sessions in einem Aufruf.

In [7]:
sys.argv = [
    "notebook",
    "--model_type", model_type,
    "--session_dir", session_dir,
    "--checkpoint_path", checkpoint_path,
    "--beam_size", str(beam_size),
    "--max_length", str(max_length), 
    "--output_dir_name", output_dir_name,

    ]
#inference.main()

Loading avsr_cocktail model...


/home/josch080/Projektgruppe/mcorec_train/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


Loading model from model-bin/avsr_cocktail_mcorec_finetune
avsr_cocktail model loaded successfully!
Inferring 25 sessions using avsr_cocktail model
Processing session session_44


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speakers:  17%|█▋        | 1/6 [01:21<06:47, 81.52s/it]


Processing speakers:  33%|███▎      | 2/6 [02:01<03:47, 56.96s/it]


Processing speakers:  50%|█████     | 3/6 [03:03<02:57, 59.32s/it]


Processing speaker spk_3 track 1 of 2: 100%|██████████| 15/15 [00:23<00:00,  1.59s/it]

Processing speakers:  67%|██████▋   | 4/6 [03:42<01:42, 51.47s/it]


Processing speakers:  83%|████████▎ | 5/6 [04:33<00:51, 51.09s/it]


Processing speaker spk_5 track 1 of 3: 100%|██████████| 23/23 [00:23<00:00,  1.01s/it]

Processing speaker spk_5 track 2 of 3: 100%|██████████| 2/2 [00:00<00:00,  2.06it/s]

Processing speakers: 100%|██████████| 6/6 [05:03<00:00, 50.55s/it]


Processing session session_57


Processing speakers:   0%|          | 0/5 [00:00<?, ?it/s]


Processing speaker spk_0 track 1 of 2: 100%|██████████| 21/21 [01:12<00:00,  3.46s/it]

Processing speakers:  20%|██        | 1/5 [01:33<06:13, 93.48s/it]


Processing speakers:  40%|████      | 2/5 [02:16<03:11, 63.90s/it]


Processing speakers:  60%|██████    | 3/5 [02:50<01:40, 50.04s/it]


Processing speaker spk_3 track 1 of 2: 100%|██████████| 21/21 [00:27<00:00,  1.31s/it]

Processing speakers:  80%|████████  | 4/5 [03:47<00:52, 53.00s/it]


Processing speakers: 100%|██████████| 5/5 [05:05<00:00, 61.18s/it]


Processing session session_139


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speakers:  17%|█▋        | 1/6 [01:24<07:01, 84.34s/it]


Processing speakers:  33%|███▎      | 2/6 [02:35<05:06, 76.54s/it]


Processing speakers:  50%|█████     | 3/6 [03:22<03:09, 63.11s/it]


Processing speaker spk_3 track 1 of 2: 100%|██████████| 19/19 [00:25<00:00,  1.36s/it]

Processing speakers:  67%|██████▋   | 4/6 [03:57<01:43, 51.99s/it]


Processing speakers:  83%|████████▎ | 5/6 [04:43<00:49, 49.93s/it]


Processing speakers: 100%|██████████| 6/6 [05:44<00:00, 57.45s/it]


Processing session session_133


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speaker spk_0 track 1 of 2: 100%|██████████| 22/22 [00:41<00:00,  1.89s/it]

Processing speakers:  17%|█▋        | 1/6 [00:53<04:29, 53.88s/it]


Processing speaker spk_1 track 1 of 2: 100%|██████████| 14/14 [00:30<00:00,  2.16s/it]

Processing speakers:  33%|███▎      | 2/6 [01:56<03:55, 58.76s/it]


Processing speakers:  50%|█████     | 3/6 [03:32<03:47, 75.90s/it]


Processing speakers:  67%|██████▋   | 4/6 [04:11<02:02, 61.47s/it]


Processing speakers:  83%|████████▎ | 5/6 [05:13<01:01, 61.56s/it]


Processing speakers: 100%|██████████| 6/6 [06:46<00:00, 67.81s/it]


Processing session session_134


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speakers:  17%|█▋        | 1/6 [01:21<06:46, 81.30s/it]


Processing speakers:  33%|███▎      | 2/6 [02:27<04:49, 72.32s/it]


Processing speakers:  50%|█████     | 3/6 [03:33<03:29, 69.70s/it]


Processing speakers:  67%|██████▋   | 4/6 [04:51<02:25, 72.62s/it]


Processing speaker spk_4 track 1 of 2: 100%|██████████| 17/17 [00:22<00:00,  1.33s/it]

Processing speakers:  83%|████████▎ | 5/6 [05:43<01:05, 65.23s/it]


Processing speakers: 100%|██████████| 6/6 [06:38<00:00, 66.38s/it]


Processing session session_50


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speakers:  17%|█▋        | 1/6 [00:38<03:14, 38.83s/it]


Processing speakers:  33%|███▎      | 2/6 [01:24<02:51, 42.93s/it]


Processing speaker spk_2 track 1 of 2: 100%|██████████| 18/18 [00:21<00:00,  1.21s/it]

Processing speakers:  50%|█████     | 3/6 [01:53<01:49, 36.42s/it]


Processing speaker spk_3 track 1 of 3: 100%|██████████| 16/16 [00:22<00:00,  1.38s/it]

Processing speaker spk_3 track 2 of 3: 100%|██████████| 11/11 [00:24<00:00,  2.22s/it]

Processing speakers:  67%|██████▋   | 4/6 [02:57<01:34, 47.35s/it]


Processing speakers:  83%|████████▎ | 5/6 [03:25<00:40, 40.48s/it]


Processing speakers: 100%|██████████| 6/6 [04:32<00:00, 45.42s/it]


Processing session session_49


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speakers:  17%|█▋        | 1/6 [00:16<01:20, 16.01s/it]


Processing speakers:  33%|███▎      | 2/6 [00:50<01:47, 26.84s/it]


Processing speaker spk_2 track 1 of 8: 100%|██████████| 1/1 [00:00<00:00,  1.60it/s]

Processing speaker spk_2 track 2 of 8: 100%|██████████| 1/1 [00:00<00:00,  1.73it/s]

Processing speaker spk_2 track 3 of 8: 100%|██████████| 1/1 [00:00<00:00,  1.98it/s]

Processing speaker spk_2 track 4 of 8: 100%|██████████| 3/3 [00:16<00:00,  5.47s/it]

Processing speaker spk_2 track 5 of 8: 100%|██████████| 2/2 [00:08<00:00,  4.41s/it]

Processing speaker spk_2 track 6 of 8: 100%|██████████| 8/8 [00:17<00:00,  2.21s/it]

Processing speaker spk_2 track 7 of 8: 100%|██████████| 3/3 [00:02<00:00,  1.06it/s]

Processing speakers:  50%|█████     | 3/6 [01:55<02:12, 44.16s/it]


Processing speakers:  67%|██████▋   | 4/6 [02:40<01:29, 44.58s/it]


Processing speakers:  83%|████████▎ | 5/6 [03:46<00:52, 52.37s/it]


Processing speaker spk_5 track 1 of 2: 100%|██████████| 21/21 [00:45<00:00,  2.15s/it]

Processing speakers: 100%|██████████| 6/6 [04:50<00:00, 48.34s/it]


Processing session session_43


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speaker spk_0 track 1 of 2: 100%|██████████| 27/27 [01:07<00:00,  2.49s/it]

Processing speakers:  17%|█▋        | 1/6 [01:27<07:18, 87.67s/it]


Processing speakers:  33%|███▎      | 2/6 [02:13<04:13, 63.25s/it]


Processing speakers:  50%|█████     | 3/6 [03:19<03:13, 64.47s/it]


Processing speakers:  67%|██████▋   | 4/6 [04:41<02:22, 71.15s/it]


Processing speaker spk_4 track 1 of 2: 100%|██████████| 16/16 [00:25<00:00,  1.61s/it]

Processing speakers:  83%|████████▎ | 5/6 [05:36<01:05, 65.63s/it]


Processing speakers: 100%|██████████| 6/6 [06:25<00:00, 64.30s/it]


Processing session session_54


Processing speakers:   0%|          | 0/5 [00:00<?, ?it/s]


Processing speaker spk_0 track 1 of 2: 100%|██████████| 26/26 [00:46<00:00,  1.81s/it]

Processing speakers:  20%|██        | 1/5 [01:01<04:05, 61.44s/it]


Processing speakers:  40%|████      | 2/5 [02:20<03:35, 71.93s/it]


Processing speakers:  60%|██████    | 3/5 [03:15<02:08, 64.20s/it]


Processing speakers:  80%|████████  | 4/5 [04:41<01:12, 72.82s/it]


Processing speakers: 100%|██████████| 5/5 [05:24<00:00, 64.99s/it]


Processing session session_140


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speakers:  17%|█▋        | 1/6 [00:50<04:13, 50.70s/it]


Processing speakers:  33%|███▎      | 2/6 [01:32<03:02, 45.51s/it]


Processing speakers:  50%|█████     | 3/6 [02:53<03:04, 61.60s/it]


Processing speaker spk_3 track 1 of 3: 100%|██████████| 5/5 [00:07<00:00,  1.44s/it]

Processing speaker spk_3 track 2 of 3: 100%|██████████| 11/11 [00:12<00:00,  1.13s/it]

Processing speakers:  67%|██████▋   | 4/6 [03:33<01:46, 53.04s/it]


Processing speakers:  83%|████████▎ | 5/6 [04:18<00:50, 50.19s/it]


Processing speaker spk_5 track 1 of 3: 100%|██████████| 6/6 [00:07<00:00,  1.33s/it]

Processing speaker spk_5 track 2 of 3: 100%|██████████| 5/5 [00:04<00:00,  1.01it/s]

Processing speakers: 100%|██████████| 6/6 [04:52<00:00, 48.70s/it]


Processing session session_137


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speakers:  17%|█▋        | 1/6 [00:36<03:03, 36.64s/it]


Processing speakers:  33%|███▎      | 2/6 [02:26<05:17, 79.45s/it]


Processing speakers:  50%|█████     | 3/6 [03:03<03:00, 60.21s/it]


Processing speaker spk_3 track 1 of 2: 100%|██████████| 20/20 [00:21<00:00,  1.09s/it]

Processing speakers:  67%|██████▋   | 4/6 [03:53<01:52, 56.18s/it]


Processing speakers:  83%|████████▎ | 5/6 [04:48<00:55, 55.72s/it]


Processing speakers: 100%|██████████| 6/6 [05:42<00:00, 57.14s/it]


Processing session session_53


Processing speakers:   0%|          | 0/5 [00:00<?, ?it/s]


Processing speakers:  20%|██        | 1/5 [01:17<05:09, 77.37s/it]


Processing speakers:  40%|████      | 2/5 [02:24<03:34, 71.49s/it]


Processing speakers:  60%|██████    | 3/5 [03:22<02:10, 65.42s/it]


Processing speaker spk_3 track 1 of 2: 100%|██████████| 20/20 [01:13<00:00,  3.70s/it]

Processing speakers:  80%|████████  | 4/5 [04:57<01:16, 76.99s/it]


Processing speakers: 100%|██████████| 5/5 [05:39<00:00, 67.84s/it]


Processing session session_40


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speakers:  17%|█▋        | 1/6 [00:42<03:34, 42.84s/it]


Processing speakers:  33%|███▎      | 2/6 [02:14<04:46, 71.53s/it]


Processing speaker spk_2 track 1 of 3: 0it [00:00, ?it/s]

Processing speaker spk_2 track 2 of 3: 100%|██████████| 13/13 [00:46<00:00,  3.60s/it]

Processing speakers:  50%|█████     | 3/6 [03:46<04:02, 80.69s/it]


Processing speaker spk_3 track 1 of 2: 100%|██████████| 18/18 [00:45<00:00,  2.51s/it]

Processing speakers:  67%|██████▋   | 4/6 [04:57<02:33, 76.98s/it]


Processing speakers:  83%|████████▎ | 5/6 [06:03<01:13, 73.04s/it]


Processing speakers: 100%|██████████| 6/6 [07:20<00:00, 73.40s/it]


Processing session session_42


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speaker spk_0 track 1 of 2: 100%|██████████| 12/12 [00:13<00:00,  1.15s/it]

Processing speakers:  17%|█▋        | 1/6 [00:50<04:12, 50.57s/it]


Processing speakers:  33%|███▎      | 2/6 [02:28<05:13, 78.33s/it]


Processing speakers:  50%|█████     | 3/6 [04:09<04:26, 88.81s/it]


Processing speaker spk_3 track 1 of 2: 100%|██████████| 4/4 [00:05<00:00,  1.33s/it]

Processing speakers:  67%|██████▋   | 4/6 [05:05<02:31, 75.73s/it]


Processing speakers:  83%|████████▎ | 5/6 [06:09<01:11, 71.62s/it]


Processing speakers: 100%|██████████| 6/6 [07:00<00:00, 70.07s/it]


Processing session session_48


Processing speakers:   0%|          | 0/3 [00:00<?, ?it/s]


Processing speaker spk_1 track 1 of 3: 100%|██████████| 2/2 [00:01<00:00,  1.40it/s]

Processing speaker spk_1 track 2 of 3: 100%|██████████| 6/6 [00:04<00:00,  1.38it/s]

Processing speakers:  33%|███▎      | 1/3 [00:54<01:48, 54.00s/it]


Processing speakers:  67%|██████▋   | 2/3 [01:25<00:40, 40.59s/it]


Processing speaker spk_3 track 1 of 10: 100%|██████████| 4/4 [00:04<00:00,  1.07s/it]

Processing speaker spk_3 track 2 of 10: 0it [00:00, ?it/s]

Processing speaker spk_3 track 3 of 10: 100%|██████████| 2/2 [00:01<00:00,  1.37it/s]

Processing speaker spk_3 track 4 of 10: 100%|██████████| 9/9 [00:17<00:00,  1.93s/it]

Processing speaker spk_3 track 5 of 10: 100%|██████████| 8/8 [00:15<00:00,  1.96s/it]

Processing speaker spk_3 track 6 of 10: 100%|██████████| 2/2 [00:01<00:00,  1.01it/s]

Processing speaker spk_3 track 7 of 10: 100%|██████████| 2/2 [00:06<00:00,  3.04s/it]

Processing speaker spk_3 track 8 of 10: 100%|██████████| 1/1 [00:08<00:00,  8.23s/it]

Processing speaker spk_3 track 9 of 10: 0it [00:00, ?it/s]

Processing speakers: 100%|██████████| 3/3 [02:28<00:00, 49.43s/it]


Processing session session_135


Processing speakers:   0%|          | 0/5 [00:00<?, ?it/s]


Processing speaker spk_0 track 1 of 2: 100%|██████████| 15/15 [00:23<00:00,  1.55s/it]

Processing speakers:  20%|██        | 1/5 [01:00<04:03, 60.90s/it]


Processing speakers:  40%|████      | 2/5 [02:03<03:06, 62.19s/it]


Processing speakers:  60%|██████    | 3/5 [03:24<02:21, 70.68s/it]


Processing speakers:  80%|████████  | 4/5 [04:25<01:06, 66.62s/it]


Processing speakers: 100%|██████████| 5/5 [05:40<00:00, 68.04s/it]


Processing session session_51


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speakers:  17%|█▋        | 1/6 [00:47<03:58, 47.75s/it]


Processing speaker spk_1 track 1 of 2: 100%|██████████| 14/14 [00:27<00:00,  1.94s/it]

Processing speakers:  33%|███▎      | 2/6 [01:52<03:50, 57.54s/it]


Processing speakers:  50%|█████     | 3/6 [02:31<02:27, 49.03s/it]


Processing speakers:  67%|██████▋   | 4/6 [04:10<02:17, 68.76s/it]


Processing speakers:  83%|████████▎ | 5/6 [04:58<01:01, 61.37s/it]


Processing speakers: 100%|██████████| 6/6 [05:32<00:00, 55.37s/it]


Processing session session_56


Processing speakers:   0%|          | 0/4 [00:00<?, ?it/s]


Processing speakers:  25%|██▌       | 1/4 [00:30<01:32, 30.69s/it]


Processing speakers:  50%|█████     | 2/4 [02:10<02:23, 71.61s/it]


Processing speakers:  75%|███████▌  | 3/4 [03:37<01:18, 78.63s/it]


Processing speakers: 100%|██████████| 4/4 [03:52<00:00, 58.07s/it]


Processing session session_132


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speakers:  17%|█▋        | 1/6 [01:38<08:12, 98.57s/it]


Processing speakers:  33%|███▎      | 2/6 [02:46<05:22, 80.63s/it]


Processing speakers:  50%|█████     | 3/6 [04:28<04:30, 90.30s/it]


Processing speakers:  67%|██████▋   | 4/6 [05:08<02:21, 70.59s/it]


Processing speakers:  83%|████████▎ | 5/6 [06:03<01:05, 65.02s/it]


Processing speaker spk_5 track 1 of 5: 100%|██████████| 10/10 [00:27<00:00,  2.72s/it]

Processing speaker spk_5 track 2 of 5: 100%|██████████| 4/4 [00:02<00:00,  1.45it/s]

Processing speaker spk_5 track 3 of 5: 100%|██████████| 11/11 [00:22<00:00,  2.01s/it]

Processing speaker spk_5 track 4 of 5: 100%|██████████| 4/4 [00:04<00:00,  1.23s/it]

Processing speakers: 100%|██████████| 6/6 [07:02<00:00, 70.36s/it]


Processing session session_138


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speakers:  17%|█▋        | 1/6 [00:29<02:25, 29.03s/it]


Processing speakers:  33%|███▎      | 2/6 [02:06<04:36, 69.17s/it]


Processing speakers:  50%|█████     | 3/6 [02:54<02:58, 59.52s/it]


Processing speaker spk_3 track 1 of 2: 100%|██████████| 21/21 [00:46<00:00,  2.20s/it]

Processing speakers:  67%|██████▋   | 4/6 [03:56<02:01, 60.62s/it]


Processing speakers:  83%|████████▎ | 5/6 [04:33<00:51, 51.97s/it]


Processing speaker spk_5 track 1 of 4: 100%|██████████| 7/7 [00:06<00:00,  1.13it/s]

Processing speaker spk_5 track 2 of 4: 100%|██████████| 15/15 [00:32<00:00,  2.15s/it]

Processing speaker spk_5 track 3 of 4: 100%|██████████| 6/6 [00:05<00:00,  1.02it/s]

Processing speakers: 100%|██████████| 6/6 [05:20<00:00, 53.44s/it]


Processing session session_41


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speaker spk_0 track 1 of 2: 100%|██████████| 19/19 [00:24<00:00,  1.28s/it]

Processing speakers:  17%|█▋        | 1/6 [00:25<02:08, 25.74s/it]


Processing speaker spk_1 track 1 of 2: 0it [00:00, ?it/s]

Processing speakers:  33%|███▎      | 2/6 [02:17<05:05, 76.31s/it]


Processing speaker spk_2 track 1 of 2: 0it [00:00, ?it/s]

Processing speakers:  50%|█████     | 3/6 [03:45<04:04, 81.54s/it]


Processing speaker spk_3 track 1 of 2: 100%|██████████| 3/3 [00:09<00:00,  3.17s/it]

Processing speakers:  67%|██████▋   | 4/6 [04:38<02:20, 70.42s/it]


Processing speakers:  83%|████████▎ | 5/6 [05:38<01:06, 66.75s/it]


Processing speaker spk_5 track 1 of 2: 100%|██████████| 29/29 [00:57<00:00,  1.99s/it]

Processing speakers: 100%|██████████| 6/6 [06:37<00:00, 66.20s/it]


Processing session session_136


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speakers:  17%|█▋        | 1/6 [01:32<07:44, 92.96s/it]


Processing speakers:  33%|███▎      | 2/6 [02:25<04:36, 69.23s/it]


Processing speakers:  50%|█████     | 3/6 [03:16<03:02, 60.86s/it]


Processing speakers:  67%|██████▋   | 4/6 [04:18<02:02, 61.47s/it]


Processing speaker spk_4 track 1 of 2: 100%|██████████| 20/20 [01:09<00:00,  3.47s/it]

Processing speakers:  83%|████████▎ | 5/6 [05:40<01:08, 68.70s/it]


Processing speakers: 100%|██████████| 6/6 [06:19<00:00, 63.33s/it]


Processing session session_141


Processing speakers:   0%|          | 0/5 [00:00<?, ?it/s]


Processing speakers:  20%|██        | 1/5 [00:47<03:08, 47.00s/it]


Processing speakers:  40%|████      | 2/5 [01:54<02:56, 58.96s/it]


Processing speakers:  60%|██████    | 3/5 [02:46<01:51, 55.79s/it]


Processing speakers:  80%|████████  | 4/5 [03:20<00:47, 47.26s/it]


Processing speakers: 100%|██████████| 5/5 [04:07<00:00, 49.41s/it]


Processing session session_52


Processing speakers:   0%|          | 0/6 [00:00<?, ?it/s]


Processing speakers:  17%|█▋        | 1/6 [00:39<03:19, 39.92s/it]


Processing speakers:  33%|███▎      | 2/6 [00:59<01:51, 27.85s/it]


Processing speaker spk_2 track 1 of 2: 100%|██████████| 2/2 [00:01<00:00,  1.36it/s]

Processing speakers:  50%|█████     | 3/6 [02:32<02:53, 57.83s/it]


Processing speakers:  67%|██████▋   | 4/6 [03:19<01:46, 53.33s/it]


Processing speakers:  83%|████████▎ | 5/6 [03:39<00:41, 41.28s/it]


Processing speakers: 100%|██████████| 6/6 [05:36<00:00, 56.10s/it]


Processing session session_55


Processing speakers:   0%|          | 0/5 [00:00<?, ?it/s]


Processing speaker spk_0 track 1 of 2: 100%|██████████| 23/23 [00:41<00:00,  1.79s/it]

Processing speakers:  20%|██        | 1/5 [01:10<04:42, 70.66s/it]


Processing speakers:  40%|████      | 2/5 [02:26<03:40, 73.63s/it]


Processing speakers:  60%|██████    | 3/5 [03:24<02:13, 66.52s/it]


Processing speakers:  80%|████████  | 4/5 [03:44<00:48, 48.05s/it]


Processing speakers: 100%|██████████| 5/5 [05:38<00:00, 67.68s/it]


## 5 – Evaluation via Subprocess

`evaluate.py` wird als separater Prozess aufgerufen, damit sein Output
vollständig gepuffert und weiterverarbeitet werden kann.
Das Ergebnis wird zusätzlich in `eval_log_final_results.txt` persistiert.

In [15]:
import subprocess, sys, re, pandas as pd

eval_session_dir = session_dir   
eval_output_prefix = output_dir_name

cmd = [
    sys.executable,
    "script/evaluate.py",
    "--session_dir", eval_session_dir,
    "--output_dir_name", eval_output_prefix,
    "--label_dir_name", "labels",
]

result = subprocess.run(cmd, capture_output=True, text=True)

# Ausgabe trotzdem im Notebook sehen:
print(result.stdout)

# Ausgabe auch in eine Datei schreiben
with open("eval_log_final_results.txt", "w") as f:
    f.write(result.stdout)


Evaluating 25 sessions

=== Evaluating session session_44 ===

--- Evaluating output dir: output_final_bs12_len20 ---
Conversation clustering F1 score: 0.6
Speaker to WER: {'spk_0': 0.4006, 'spk_1': 0.6334, 'spk_2': 0.538, 'spk_3': 0.611, 'spk_4': 0.6278, 'spk_5': 0.6301}
Speaker clustering F1 score: {'spk_0': 1.0, 'spk_1': 1.0, 'spk_2': 0.5, 'spk_3': 0.5, 'spk_4': 0.5, 'spk_5': 0.5}
Joint ASR-Clustering Error Rate: {'spk_0': 0.2003, 'spk_1': 0.3167, 'spk_2': 0.519, 'spk_3': 0.5555, 'spk_4': 0.5639000000000001, 'spk_5': 0.56505}

=== Evaluating session session_57 ===

--- Evaluating output dir: output_final_bs12_len20 ---
Conversation clustering F1 score: 1.0
Speaker to WER: {'spk_0': 0.335, 'spk_1': 0.3285, 'spk_2': 0.4651, 'spk_3': 0.5205, 'spk_4': 0.5283}
Speaker clustering F1 score: {'spk_0': 1.0, 'spk_1': 1.0, 'spk_2': 1.0, 'spk_3': 1.0, 'spk_4': 1.0}
Joint ASR-Clustering Error Rate: {'spk_0': 0.1675, 'spk_1': 0.16425, 'spk_2': 0.23255, 'spk_3': 0.26025, 'spk_4': 0.26415}

=== Eva

## 6 – Ergebnis-Parsing

Regex extrahiert F1, WER und Joint Error aus dem `evaluate.py`-Output.
Das DataFrame `baseline_results` enthält eine Zeile pro Session-Experiment-Kombination.

In [16]:
# Regex: matcht einen Ergebnis-Block aus dem evaluate.py-Output
pattern = re.compile(
    r"Results for output dir prefix variant '([^']+)':\s+"
    r"\s*Average Conversation Clustering F1 score: ([0-9.]+)\s+"
    r"\s*Average Speaker WER: ([0-9.]+)\s+"
    r"\s*Average Joint ASR-Clustering Error Rate: ([0-9.]+)",
    re.MULTILINE
)

rows = []
for m in pattern.finditer(result.stdout):
    rows.append({
        "exp": m.group(1),                 # z.B. "output_avsr_cocktail_finetuned"
        "avg_conv_f1": float(m.group(2)),
        "avg_speaker_wer": float(m.group(3)),
        "avg_joint_error": float(m.group(4)),
    })

baseline_results = pd.DataFrame(rows)
# 'output_'-Prefix aus exp-Namen entfernen für lesbarere model-Spalte
baseline_results["model"] = baseline_results["exp"].str.replace("^output_", "", regex=True)
baseline_results

exp  avg_conv_f1  avg_speaker_wer  avg_joint_error  \
0  output_final_bs12_len20     0.815282         0.497967         0.354294   

              model  
0  final_bs12_len20

In [17]:
baseline_results.to_csv(
    "final_results.csv",
    index=False
)


## 7 – Vergleich: Final vs. BL4-Baseline

In [14]:
import pandas as pd

final_df = baseline_results.copy()   # Ergebnisse aus Abschnitt 6 (beam=12, len=20)

base_df = pd.read_csv("results_baseline_dev_without_central_videos.csv")

BASELINE_EXP = "output_avsr_cocktail_finetuned"   # BL4 Baseline: beam=3, len=15
FINAL_EXP = "output_final_bs12_len20"

def get_row(df, exp_name):
    if "exp" not in df.columns:
        raise ValueError(f"Keine Spalte 'exp' gefunden. Spalten: {list(df.columns)}")
    hit = df[df["exp"].astype(str) == exp_name]
    if hit.empty:
        print(f"[WARN] exp='{exp_name}' nicht gefunden. Verfügbare exp: {df['exp'].tolist()}")
        return df.iloc[0]
    return hit.iloc[0]

b = get_row(base_df, BASELINE_EXP)
f = get_row(final_df, FINAL_EXP)

# Runs-Tabelle (zwei Zeilen: Baseline + Final)
runs = pd.DataFrame([
    {"run": "baseline (finetuned, beam=3,len=15)", "exp": b["exp"],
     "avg_conv_f1": float(b["avg_conv_f1"]),
     "avg_speaker_wer": float(b["avg_speaker_wer"]),
     "avg_joint_error": float(b["avg_joint_error"])},
    {"run": "final (beam=12,len=20)", "exp": f["exp"],
     "avg_conv_f1": float(f["avg_conv_f1"]),
     "avg_speaker_wer": float(f["avg_speaker_wer"]),
     "avg_joint_error": float(f["avg_joint_error"])},
])

display(runs.round(6))

# Vergleichstabelle: absolute und relative Verbesserung
# F1 ↑ besser (improvement = final - baseline)
# WER/JER ↓ besser (improvement = baseline - final)
comparison = pd.DataFrame([
    {
        "metric": "Avg Conv Clustering F1 (↑)",
        "baseline": runs.loc[0, "avg_conv_f1"],
        "final": runs.loc[1, "avg_conv_f1"],
        "improvement_abs": runs.loc[1, "avg_conv_f1"] - runs.loc[0, "avg_conv_f1"],
    },
    {
        "metric": "Avg Speaker WER (↓)",
        "baseline": runs.loc[0, "avg_speaker_wer"],
        "final": runs.loc[1, "avg_speaker_wer"],
        "improvement_abs": runs.loc[0, "avg_speaker_wer"] - runs.loc[1, "avg_speaker_wer"],
    },
    {
        "metric": "Avg Joint Error (↓)",
        "baseline": runs.loc[0, "avg_joint_error"],
        "final": runs.loc[1, "avg_joint_error"],
        "improvement_abs": runs.loc[0, "avg_joint_error"] - runs.loc[1, "avg_joint_error"],
    },
])
comparison["improvement_rel_%"] = (comparison["improvement_abs"] / comparison["baseline"]) * 100

display(comparison.round(6))


run                             exp  \
0  baseline (finetuned, beam=3,len=15)  output_avsr_cocktail_finetuned   
1               final (beam=12,len=20)         output_final_bs12_len20   

   avg_conv_f1  avg_speaker_wer  avg_joint_error  
0     0.815282         0.498727         0.354674  
1     0.815282         0.497967         0.354294

metric  baseline     final  improvement_abs  \
0  Avg Conv Clustering F1 (↑)  0.815282  0.815282          0.00000   
1         Avg Speaker WER (↓)  0.498727  0.497967          0.00076   
2         Avg Joint Error (↓)  0.354674  0.354294          0.00038   

   improvement_rel_%  
0           0.000000  
1           0.152474  
2           0.107202

## 8 – Interpretation

Die Parameteroptimierung (beam=12, len=20) auf **allen Dev-Sessions** bestätigt
das Ergebnis vom 5-Session-Subset:

| Metrik | Baseline (beam=3, len=15) | Final (beam=12, len=20) | Δ abs | Δ rel |
|--------|--------------------------|------------------------|-------|-------|
| Speaker WER (↓) | 0.4987 | 0.4980 | −0.0008 | −0.15% |
| Joint Error (↓) | 0.3457 | 0.3453 | −0.0004 | −0.11% |
| Conv F1 (↑) | unverändert | unverändert | 0 | 0% |

Die Verbesserungen sind **marginal aber konsistent**.
Weitere Verbesserungen folgen durch den Segmentierungs-Bugfix + min_duration-Grid (`03b_`, `02k_`, `02l_`).
